### ISSUE TITLE
Missed blocks if they are mined before lnd connects to backend

### ISSUE THREAD

**Issue by ripper234** [lightningnetwork/lnd#1188]
### Background

Following the tutorial, I sent a funding transaction from Alice to create the Alice <-> Bob channel.
The transaction is kept on pending, and I'm seeing a lot of out of order blocks on the logs.

### Your environment

* lnd version 0.4.1-beta commit=097fd50747356a39593e5ac7625bf0b0fb1305a1
* Windows 10
* btcd compiled from source today from roasbeef branch commit a03db407e40d3b66ea29984263bbc8bf4d2f04c4

### Steps to reproduce

Just following the tutorial. After generating coins on Alice's and Charlie's addresses, I changed the mining address to another unrelated address, not sure if that's related.

>lncli --rpcserver=localhost:10001 --no-macaroons openchannel -node_key=0365dccab6bb2df9a9d2cce7b88be76a5c572df64931fa18012a4a20fb274c71c1 1000000
{
        "funding_txid": "9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b"
}

C:\work\lnd>btcctl generate 6
[
  "417f91cce62d9302ed722030375674e55d8751ae823a9b3f4ab7946f21f16e4b",
  "43aed7ec8d1a025fccdf0c53704a1f50c56a114e1414a695527c29c99bfe8eb3",
  "2357b2e4ec06c0d27426e085ae280bb480efd9c4bffac8405ab1c635a28c6ab5",
  "4539668e0220e9c33a693fc43b63c9e78819570c66c0148c76969a0a6c71a5fa",
  "5918235ddc5cb6511ae37c4fcbc63007f7ac0036d35ea5353b5207dadfcc2c82",
  "780b49dbbcbbd83d8c4ea830d77c1f9776cdcb86a363bb4d7a4f0acc34e55d4b"
]

>lncli --rpcserver=localhost:10001 --no-macaroons listchannels
{
    "channels": [
    ]
}

>lncli --rpcserver=localhost:10001 --no-macaroons pendingchannels
{
    "total_limbo_balance": "0",
    "pending_open_channels": [
        {
            "channel": {
                "remote_node_pub": "0365dccab6bb2df9a9d2cce7b88be76a5c572df64931fa18012a4a20fb274c71c1",
                "channel_point": "9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b:0",
                "capacity": "1000000",
                "local_balance": "990950",
                "remote_balance": "0"
            },
            "confirmation_height": 0,
            "commit_fee": "9050",
            "commit_weight": "600",
            "fee_per_kw": "12500"
        }
    ],
    "pending_closing_channels": [
    ],
    "pending_force_closing_channels": [
    ],
    "waiting_close_channels": [
    ]
}

Full lnd logs on pastebin:

https://pastebin.com/e2qJ2gp5

> 2018-05-06 10:47:01.122 [INF] NTFN: New block epoch subscription
> 2018-05-06 10:47:01.124 [ERR] LNWL: Notifying unmined tx notification (9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b) while creating notification for blocks
> 2018-05-06 10:47:01.136 [INF] CNCT: ChannelArbitrator(9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b:0): starting state=StateDefault
> 2018-05-06 10:47:01.142 [DBG] CNCT: ChannelArbitrator(9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b:0): new block (height=500) examining active HTLC's
> 2018-05-06 10:47:01.142 [DBG] CNCT: ChannelArbitrator(9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b:0): checking chain actions at height=500
> 2018-05-06 10:47:01.143 [DBG] CNCT: Starting chain watcher for ChannelPoint(9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b:0)
> 2018-05-06 10:47:01.144 [INF] NTFN: New spend subscription: utxo=9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b:0
> 2018-05-06 10:47:01.145 [INF] FNDG: Finalizing pendingID(4902526390187e9d101381661db6587194267d48a7c10e7a20ed4b3bf21c70bb) over ChannelPoint(9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b:0), waiting for channel open on-chain
> 2018-05-06 10:47:01.145 [INF] CNCT: Close observer for ChannelPoint(9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b:0) active
> 2018-05-06 10:47:01.146 [INF] FNDG: Waiting for funding tx (9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b) to reach 3 confirmations
> 2018-05-06 10:47:01.146 [INF] NTFN: New confirmation subscription: txid=9080f300600382f79e75bd2997c1ccf4beac7fc9ee75db75bf1c7f5369bbb66b, numconfs=3
> 2018-05-06 10:47:15.474 [DBG] PEER: Sending Ping to 127.0.0.1:10012
> 2018-05-06 10:47:15.474 [DBG] PEER: Received Ping from 127.0.0.1:10012
> 2018-05-06 10:47:15.479 [DBG] PEER: Sending Pong to 127.0.0.1:10012
> 2018-05-06 10:47:15.482 [DBG] PEER: Received Pong from 127.0.0.1:10012
> 2018-05-06 10:47:16.974 [ERR] CRTR: out of order block: expecting height=401, got height=501
> 2018-05-06 10:47:16.974 [WRN] NTFN: Received blocks out of order: current height=400, new height=501
> 2018-05-06 10:47:16.987 [WRN] NTFN: Received blocks out of order: current height=400, new height=502

**Comment by ripper234** [comment]
Restarting lnd & btcd seems to resolve the issue. I'm keeping it open because it shouldn't? haven't happened in the first place.

Rest of the log after restart: https://pastebin.com/3DGb7qu8

**Comment by Roasbeef** [comment]
Did you mine enough blocks for segwit to activate before starting your scenario? 

> Just following the tutorial. After generating coins on Alice's and Charlie's addresses, I changed the mining address to another unrelated address, not sure if that's related.

In this case, since you had to restart btcd, you'll need to wait for the `lnd` nodes to reconnect properly. Otherwise, you may start mining blocks w/o them first establishing a connection to the RPC backend.

**Comment by ripper234** [comment]
> Did you mine enough blocks for segwit to activate before starting your scenario?

Yes, mined 400+ blocks before.

> > Just following the tutorial. After generating coins on Alice's and Charlie's addresses, I changed the mining address to another unrelated address, not sure if that's related.
> 
> In this case, since you had to restart btcd, you'll need to wait for the lnd nodes to reconnect properly. 
Otherwise, you may start mining blocks w/o them first establishing a connection to the RPC backend.

I don't understand the implications of this ...

**Comment by halseth** [comment]
> I don't understand the implications of this ...
Just to make sure `lnd` is ready to receive notifications about mined blocks when they get mined.

Looks like `lnd` missed about 100  blocks:
```
2018-05-06 10:47:16.974 [ERR] CRTR: out of order block: expecting height=401, got height=501
2018-05-06 10:47:16.974 [WRN] NTFN: Received blocks out of order: current height=400, new height=501
2018-05-06 10:47:16.987 [WRN] NTFN: Received blocks out of order: current height=400, new height=502
```

Not exactly sure why that happened though... @aakselrod?

**Comment by Roasbeef** [comment]
> I don't understand the implications of this ...

What I mean is that if you shutdown `btcd`, then we'll loose the connection. There'll be a gap of time where we haven't yet reconnected.

**Comment by ripper234** [comment]
So I restart btcd, the connection is lost. I generate some new blocks, and lnd doesn't seem them?
Isn't this by definition a synchronization bug? (If I understand correctly)

**Comment by ripper234** [comment]
I don't understand.

Isn't `lnd missing blocks`, for whatever reason, a bug?
Ideally shouldn't it catch up to old blocks? Or is it that it _can't_ really do that?

**Comment by Roasbeef** [comment]
On restart lnd will catch up old blocks, however it won't atm if the node
backend (which has an uptime assumption) is restarted.

On May 7, 2018 1:43 PM, "Ron Gross" <notifications@github.com> wrote:

I don't understand.

Isn't lnd missing blocks, for whatever reason, a bug?
Ideally shouldn't it catch up to old blocks? Or is it that it *can't*
really do that?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/lightningnetwork/lnd/issues/1188#issuecomment-387199965>,
or mute the thread

**Comment by ripper234** [comment]
Does the uptime assumption make sense in a decentralized environment where any service can fall/be unreachable at any time? What about a power outage that forces both lnd and the backend to restart at the same time?

**Comment by Roasbeef** [comment]
> in a decentralized environment where any service 

Well if you're connecting directly to a full node, the assumption is that the node itself is reliable. If it isn't, then you have other issues as you may be missing on chain channel events. 

> What about a power outage that forces both lnd and the backend to restart at the same time?

That would be handled properly.

**Comment by ripper234** [comment]
> Well if you're connecting directly to a full node, the assumption is that the node itself is reliable. If it isn't, then you have other issues as you may be missing on chain channel events.

What's the reliability requirement? If your full node goes down for an hour, would that break anything?

**Comment by Roasbeef** [comment]
> What's the reliability requirement? If your full node goes down for an hour, would that break anything?

It depends on the CSV values of your channels. Most values are multi-day, so your node breaking down for an hour shouldn't be an issue. It's also possible to manually set the CSV value of each channel you have in order to tune to either your risk tolerance, or an allowance for downtime.

**Comment by ripper234** [comment]
> shouldn't be an issue

So why on simnet it's an issue if btcd/lnd are restarted?

**Comment by Roasbeef** [comment]
Simnet has noting to do with it. It's that you restarted the full node and
mined blocks before lnd reconnected. Please careful read my prior comments.

**Comment by Roasbeef** [comment]
Will be fixed by #1220

**Comment by Roasbeef** [comment]
#1220 has been merged, closing.
